RFC32: CKB VM version selection (#238)
nervosnetwork/rfcs: rfcs/0032-ckb-vm-version-selection/0032-ckb-vm-version-selection.md

---
Number: "0032"
Category: Consensus (Hard Fork)
Status: Draft
Author: Ian Yang
Organization: Nervos Foundation
Created: 2021-04-26
---

# CKB VM Version Selection

## Abstract

> This RFC proposes a mechanism to choose the CKB VM version after the hard fork. What's included in the new CKB VM version is beyond the scope of this document.

Reviewer: Recommend adding a link to the VM version 1 RFC as an example of a VM version

Author: ```suggestion
This RFC proposes a mechanism to choose the CKB VM version after the hard fork. This document will not cover what's included in the new CKB VM version, please refer to [rfc33] and [rfc34] for the changes.
```

## Motivation

It's essential to keep improving CKB VM because it is the computation bottleneck of the whole network. The upgrade packages can improve the performance, bring bug fixings and add new RISC-V extensions. However the upgrade should not break the old code, users must have the opt-in option to specify the VM version.

This RFC proposes a general mechanism that determines how the CKB node chooses the CKB VM version for a transaction script group.

## Specification

> The CKB VM upgrades via hard fork. The next scheduled hard fork is ckb2021, which is activated since a specific epoch. For all the transactions in the blocks before the activation epoch, they must run the CKB VM version 0 to verify all the script groups. In these transactions, the `hash_type` in cell lock and type script must be either 0 or 1 in the serialized molecule data.

Reviewer: There is a mention of `CKB VM version 0` out of nowhere, I do recommend we put up a short section at the start of this specification, mentioning that the `Lina` fork version uses CKB VM version 0, while the new ckb2021 fork introduces a new CKB VM version 1.

Reviewer: Also it's worth mentioning that the selection of CKB VM versions, is enabled via `hash_type`. Otherwise one can get confused why this paragraph mentions `hash_type` at all, which seems to be completely irrelavant to CKB VM versions.

After ckb2021 is activated, CKB nodes must choose the CKB VM version for each script group. The allowed values for the `hash_type` field in the lock and type script are 0, 1, and 2. Cells are sorted into different groups if they have different `hash_type`. According to the value of `hash_type`:

* When the `hash_type` is 0, the script group matches code via data hash and will run the code using the CKB VM version 0.
* When the `hash_type` is 1, the script group matches code via type script hash and will run the code using the CKB VM version 1.
* When the `hash_type` is 2, the script group matches code via data hash and will run the code using the CKB VM version 1.

> | `hash_type` | matches by       | VM version |

Reviewer: I would recommend a short section breaking down the bits of `hash_type`. People will wonder why `0` in hash type means `data hash + VM version 0`, but `2` in hash type means `data hash + VM version 1`(why not use 1 here).

Author: I'll explain why we choose such encoding pattern.

However, I will not mention the bits level encoding of `hash_type` because:

1. We have not decided yet how to encode `data hash + VM version 2` and following versions
2. We want developers use this field as an enum, instead of a continuous version number plus a type flag.

| ----------- | ---------------- | ---------- |
| 0           | data hash        | 0          |
| 1           | type script hash | 1          |
| 2           | data hash        | 1          |

The VM version 0 deployed with the Lina and Aggron genesis block is version 0.

The VM version 1 will be deployed along ckb2021. The [rfc33] highlights the changes in this version, and [rfc34] lists the new syscalls only available since version 1.

[rfc33]: ../0033-ckb-vm-version-1/0033-ckb-vm-version-1.md
[rfc34]: ../0034-vm-syscalls-2/0034-vm-syscalls-2.md

The transaction is considered invalid if any `hash_type` is not in the allowed values 0, 1, and 2.

See more information about code locating using `hash_type` in [rfc22].

[rfc22]: ../0022-transaction-structure/0022-transaction-structure.md

> Because the VM selection algorithm depends on which epoch the transaction belongs to, it is not deterministic for transactions still in the memory pool. The CKB node must run two versions of transaction relay protocols, one for the current CKB version C and another for the next fork version N.

Reviewer: I feel like this is implementation details. We could add this in an appendix, but not here in the specification.

* Before the fork is activated, CKB node must relay transactions via relay protocol C and must drop all messages received via protocol N.
* After the fork is activated, CKB node must relay transactions via relay protocol N and must drop all messages received via protocol C.

The relay protocol C will be dropped after the fork succeeds. See [rfc35] for details.

[rfc35]: ../0035-ckb2021-p2p-protocol-upgrade/0035-ckb2021-p2p-protocol-upgrade.md

> Attention that, [rfc33], [rfc34], and [rfc35] must be activated together with this one starting at the same epoch.

Reviewer: Maybe rephrase this as dependencies among RFCs?

When a new block is appended to the chain and the fork is activated, or a block is rolled back and the fork is deactivated, the CKB node must rerun the verification on all the transactions in the pool.

In [nervosnetwork/ckb](https://github.com/nervosnetwork/ckb), the `hash_type` is returned in the JSON RPC as an enum. Now it has three allowed values:

* 0: "data"
* 1: "type"
* 2: "data1"

## Rationale

There are many other solutions to select VM versions. The current solution is the result of discussion and trade-off. Following are some example alternatives:

> * Always use the latest VM version. The users have no options to freeze the VM versions used in their transactions.

Reviewer 2: ```suggestion
Consistently uses the latest VM version. The users cannot specify the VM versions for transactions, and the version selection will be non-determine cause it will depend on the chain state.
```

> * Depend on the script code cell epoch. Use the old VM version if the code cell is deployed before the fork, and use the new one otherwise. The problem of this solution is that anyone can re-deploy the cell and construct the transaction using the new code cell to choose VM versions.

Reviewer 2: ```suggestion
Depend on the script code cell deployed epoch. Use the old VM version if the code cell is deployed before the fork, and use the new one otherwise. The drawback of this proposal is that anyone can re-deploy the script code after the fork, then the transaction using the newly deployed code cell will choose new VM versions. This may be unexpected behavior.
```

---
Number: "0032"
Category: Consensus (Hard Fork)
Status: Draft
Author: Ian Yang
Organization: Nervos Foundation
Created: 2021-04-26
---

> # CMB VM Version Selection

Reviewer 3: ```suggestion
# CKB VM Version Selection
```

## Abstract

This RFC proposes a mechanism to choose the CKB VM version after the hard fork. What's included in the new CKB VM version is beyond the scope of this document.

## Motivation

It's essential to keep improving CKB VM because it is the computation bottleneck of the whole network. The upgrade packages can improve the performance, bring bug fixings and add new RISC-V extensions. However the upgrade should not break the old code, users must have the opt-in option to specify the VM version.

This RFC proposes a general mechanism that determines how the CKB node chooses the CKB VM version for a transaction script group.

## Specification

The CKB VM upgrades via hard fork. The next scheduled hard fork is ckb2021, which is activated since a specific epoch. For all the transactions in the blocks before the activation epoch, they must run the CKB VM version 0 to verify all the script groups. In these transactions, the `hash_type` in cell lock and type script must be either 0 or 1 in the serialized molecule data.

> After ckb2021 is activated, CKB nodes must choose the CKB VM version for each script group. The allowed values for the `hash_type` field in the lock and type script are 0, 1, and 2. Cells are sorted into different groups if they have different `hash_type`. According to the value of `hash_type`:

Author: TODO: add cross reference between this RFC and the transaction structure RFC.

Author: Resolved by dd5f26a02d3db1659f86c42bdbf52c49975a68bb

| 1           | type script hash | 1          |
| 2           | data hash        | 1          |

See following sections about VM versions.

The transaction is considered invalid if any `hash_type` is not in the allowed values 0, 1, and 2.

Because the VM selection algorithm depends on which epoch the transaction belongs to, it is not deterministic for transactions still in the memory pool. The CKB node must run two versions of transaction relay protocols, one for the current CKB version C and another for the next fork version N.

* Before the fork is activated, CKB node must relay transactions via relay protocol C and must drop all messages received via protocol N.
* After the fork is activated, CKB node must relay transactions via relay protocol N and must drop all messages received via protocol C.

The relay protocol C will be dropped after the fork succeeds. See [rfc35] for details.

[rfc35]: ../0035-ckb2021-p2p-protocol-upgrade/0035-ckb2021-p2p-protocol-upgrade.md

When a new block is appended to the chain and the fork is activated, or a block is rolled back and the fork is deactivated, the CKB node must rerun the verification on all the transactions in the pool.

> In the [Rust implementation of CKB](https://github.com/nervosnetwork/ckb) provided by Nervos Foundation, the `hash_type` is returned as an enum. Now it has three allowed values:

Author: ```suggestion
In [nervosnetwork/ckb](https://github.com/nervosnetwork/ckb), the `hash_type` is returned in the JSON RPC as an enum. Now it has three allowed values:
```

* 0: "data"
* 1: "type"
* 2: "data1"

### CKB VM Versions

The VM version 0 deployed with the Lina and Aggron genesis block is version 0.

The VM version 1 will be deployed along ckb2021. The [rfc33] highlights the changes in this version, and [rfc34] lists the new syscalls only available since version 1.

[rfc33]: ../0033-ckb-vm-version-1/0033-ckb-vm-version-1.md

> [rfc34]: ../0034-vm-syscalls-2/0034-vm-syscalls-2.md

Reviewer 3: RFC-0035 should be listed here, too.
Those changes are all due to this RFC.

RFC-0032, RFC-0033, RFC-0034 and RFC-0035 have to be enabled together.

Author: Resolved by dd5f26a02d3db1659f86c42bdbf52c49975a68bb
